docs/extensions/payment-processors/create.md

```diff
@@ -1,5 +1,227 @@
+
+### Multiple values, e.g. changing amounts
+
+All the getters and setters take an optional extra parameter called `$label`. This can be used to store two (or more) different versions of a property, e.g. 'old' and 'new'
+
+```php
+<?php
+use Civi\Payment\PropertyBag;
+//...
+$propertyBag = new PropertyBag();
+$propertyBag->setAmount(1.23, 'old');
+$propertyBag->setAmount(2.46, 'new');
+//...
+$propertyBag->getAmount('old'); // 1.23
+$propertyBag->getAmount('new'); // 2.46
+$propertyBag->getAmount(); // throws BadMethodCall
+```
+
+This means the value is still validated and type-cast as an amount (in this example).
+
+### Custom payment processor-specific data
+
+!!! danger "todo"
+    This is not yet implemented in core.
+
+Sometimes a payment processor will requrie custom data. e.g. Stripe may use a `paymentMethodID` and a `paymentIntentID` on its `doPayment()` method.
+
+Property names should be prefixed, e.g. `stripe_paymentMethodID` and set using `PropertyBag->setCustomProperty($prop, $value, $label = 'default')`.
+
+The payment class is responsible for validating such data; anything is allowed by `setCustomProperty`, including `NULL`.
+
+Where support for custom properties is needed for a method, e.g. `doPayment()`, you should implement a method as follows:
+
+```php
+<?php
+use Civi\Payment\PropertyBag;
+use Civi\Payment\Exception\PaymentProcessorException;
+
+class CRM_Core_Payment_MyExtension extends CRM_Core_Payment {
+...
+  /**
+   * Copy our custom properties from the form data.
+   *
+   * @param PropertyBag
+   * @param Array form values
+   * @param String $component as for doPayment()
+   */
+  public function extractCustomPropertiesForDoPayment(PropertyBag $propertyBag, Array $input, $component = 'contribute') {
+
+    // We require a paymentIntentID
+    if (empty($input['paymentIntentID'])) {
+      throw new PaymentProcessorException("paymentIntentID missing");
+    }
+    // (possible validation of the value goes here since it could be malicious)
+    $propertyBag->setCustomProperty( 'stripe_paymentIntentID', $input['paymentIntentID']);
+
+    // We might need a paymentMethodID but sometimes not.
+    if (!empty($input['paymentMethodID'])) {
+      $propertyBag->setCustomProperty( 'stripe_paymentMethodID', $input['paymentMethodID']);
+    }
+  }
+  ...
+}
+```
+
+
+## The `doPayment()` method
+
+`doPayment()` will be called with the following data on the `PropertyBag` in `$params`:
+
+- `contactID`
+- `contributionID` (See [dev/financial#53](https://lab.civicrm.org/dev/financial/issues/53))
+- `amount` e.g. *1.23*
+- `isRecur`
+- `email` A billing address *foo@example.com* that should be used by the external processor (if it uses it at all)
+
+!!! danger
+    The following is a proposal
+
+
+I propose that the following changes are made throughout core: old: `$paymentProcessor->doPayment($params);` new:
```

Comment:
@eileenmcnaughton once you're back in February I think this is a place where you should weigh in.

Reply:
I don't think proposals belong in docs content.